crates/sui-kvstore/src/lib.rs

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+mod bigtable;
+use anyhow::Result;
+use async_trait::async_trait;
+pub use bigtable::client::BigTableClient;
+use sui_types::base_types::SequenceNumber;
+use sui_types::digests::TransactionDigest;
+use sui_types::effects::{TransactionEffects, TransactionEvents};
+use sui_types::full_checkpoint_content::CheckpointData;
+use sui_types::messages_checkpoint::{
+    CheckpointContents, CheckpointSequenceNumber, CheckpointSignatureMessage, CheckpointSummary,
+};
+use sui_types::object::Object;
+use sui_types::storage::ObjectKey;
+use sui_types::transaction::Transaction;
+
+#[async_trait]
+pub trait KeyValueStoreReader {
+    async fn get_objects(&mut self, objects: &[ObjectKey]) -> Result<Vec<Object>>;
+    async fn get_transactions(
+        &mut self,
+        transactions: &[TransactionDigest],
+    ) -> Result<Vec<TransactionData>>;
+    async fn get_checkpoint(
+        &mut self,
+        sequence_number: SequenceNumber,
+    ) -> Result<Option<Checkpoint>>;
```

Comment:
none of the APIs return an error on a miss.
Missing entries are not present in the result, but each entry has a method to identify the corresponding key.
(i.e. for objects: object.id(), object.version(), for transactions: tx.transaction.digest())

Reply:
Can we make this API a multi-get also?

Reply:
sure. The reason I initially special-cased it is that it's potentially the most expensive call, and after reviewing existing JSON RPC cases, looks like we don't use multi-get for checkpoints much. The exception is the sui_getCheckpoints API, but even there, it doesn't include the full checkpoints content.
Do you see a need for multi-get of checkpoints?

Reply:
In GraphQL, we would probably implement checkpoint pagination using a multi-get (no more than 50 checkpoints at a time), do you think that would be too much?

Generally, every point query that we expose through GraphQL would be implemented through a DataLoader which batches up the point queries and issues a multi-get, so that's what caused me to ask. We can configure the batch size of that loader though.

Reply:
I also think its probably best to have the API return a `Result<Vec<Option<T>>>` or `Result<Vec<Result<T>>` or something to allow for one of the individual queries to not return anything while letting the rest of the succeed and letting the caller handle how they want to deal with the partial success case

Reply:
added multi-get support for checkpoints. We can think separately about reasonable limits and how to protect the service from abuse